Add reasoning for `I18n.with_locale` (#34911)
rails/rails: guides/source/i18n.md

### Managing the Locale across Requests

A localized application will likely need to provide support for multiple locales. To accomplish this, the locale should be set at the beginning of each request so that all strings are translated using the desired locale during the lifetime of that request.

> The default locale is used for all translations unless `I18n.locale =` or `I18n.with_locale` is used.

Reviewer: ```suggestion
The default locale is used for all translations unless `I18n.locale=` or `I18n.with_locale` is used.
```

> `I18n.locale` can leak into subsequent requests served by the same thread/process if it is not consistently set in every controller. For example executing `I18n.locale = :es` in one POST requests will have effects for all later requests but only in that particular thread/process. When using a multithread web server, such as the default Puma, it might be that following multiple GET requests will have different localizations (`:en` or `:es` depending from which thread they are served). For that reason, instead of `I18n.locale =` you can use `I18n.with_locale` which does not have this leak issue.

Reviewer: I'd not even talk about web servers. This also happens with non-threaded server like unicorn. Each process will have a different locale.

```suggestion
`I18n.locale` can leak into subsequent requests served by the same thread/process if it is not consistently set in every controller. For example executing `I18n.locale = :es` in one POST requests will have effects for all later requests to controllers that don't set the locale, but only in that particular thread/process. For that reason, instead of `I18n.locale =` you can use `I18n.with_locale` which does not have this leak issue.
```
